sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -438,16 +439,27 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
       val res = sc.runJob(childRDD,
         (it: Iterator[Array[Byte]]) => if (it.hasNext) it.next() else Array.empty[Byte], p)
 
-      buf ++= res.flatMap(decodeUnsafeRows)
-
+      val partResIterator = res.iterator
```

Comment:
a bit better for perf:
```
var i = 0
while (scannedRowCount < n && i < res.length) {
  val rows = decodeUnsafeRows(res(i))
  while (scannedRowCount < n && rows.hasNext) {
    buf(scannedRowCount) = rows.next()
    scannedRowCount += 1
  }
  i += 1
}
```

It uses array indexing instead of iterator, and it inlines the method.